core/src/test/java/com/linecorp/armeria/client/limit/AsyncConcurrencyLimitTest.java

```diff
@@ -0,0 +1,151 @@
+        });
+
+        semaphore.acquire(ctx);
+        await().untilAsserted(() -> {
+            assertThat(semaphore.acquiredPermitCount()).isEqualTo(2);
+            assertThat(semaphore.availablePermitCount()).isEqualTo(0);
+        });
+    }
+
+    @Test
+    void itShouldExecuteDeferredComputationsWhenPermitsAreReleased(AsyncConcurrencyLimit sem) {
+        final CountingSemaphore semaphore = new CountingSemaphore(sem);
+
+        assertThat(semaphore.availablePermitCount()).isEqualTo(2);
+
+        semaphore.acquire(ctx);
+        semaphore.acquire(ctx);
+        semaphore.acquire(ctx);
+        semaphore.acquire(ctx);
+
+        await().untilAsserted(() -> {
+            assertThat(semaphore.count.get()).isEqualTo(2);
+            assertThat(semaphore.acquiredPermitCount()).isEqualTo(2);
+        });
+
+        semaphore.permits.poll().close();
+
+        await().untilAsserted(() -> {
+            assertThat(semaphore.count.get()).isEqualTo(3);
+            assertThat(semaphore.acquiredPermitCount()).isEqualTo(2);
+        });
+
+        semaphore.permits.poll().close();
+
+        await().untilAsserted(() -> {
+            assertThat(semaphore.count.get()).isEqualTo(4);
+            assertThat(semaphore.acquiredPermitCount()).isEqualTo(2);
+        });
+    }
+
+    private static final class CountingSemaphore {
+        private final AsyncConcurrencyLimit sem;
+        private final AtomicInteger count = new AtomicInteger(0);
+        private final ConcurrentLinkedQueue<SafeCloseable> permits = new ConcurrentLinkedQueue<>();
+
+        private CountingSemaphore(AsyncConcurrencyLimit sem) {
+            this.sem = sem;
+        }
+
+        public int availablePermitCount() {
+            return sem.availablePermitCount();
+        }
+
+        public CompletableFuture<SafeCloseable> acquire(ClientRequestContext ctx) {
+            final CompletableFuture<SafeCloseable> fPermit = sem.acquire(ctx);
+            fPermit.whenComplete((permit, throwable) -> {
+                if (throwable == null) {
+                    count.incrementAndGet();
+                    permits.offer(permit);
+                }
+            });
+            return fPermit;
+        }
+
+        public int acquiredPermitCount() {
+            return sem.acquiredPermitCount();
+        }
+    }
+
+    static class ConcurrencyLimitParameterResolver implements ParameterResolver {
+        @Override
+        public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext)
+                throws ParameterResolutionException {
+            return parameterContext.getParameter().getType() == AsyncConcurrencyLimit.class;
+        }
+
+        @Override
+        public Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext)
+                throws ParameterResolutionException {
+            return new AsyncConcurrencyLimit(100000, 2, 100);
```

Comment:
How about just creating this `AsyncConcurrencyLimit` in the method instead of using `ParameterResolver`? because we create only one instance and this `ParameterResolver` distracts reviewers' attention. 😉